pkg/ipcache/ipcache.go

```diff
@@ -0,0 +1,213 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ipcache
+
+import (
+	"encoding/json"
+	"path"
+	"sort"
+	"sync"
+
+	"github.com/cilium/cilium/pkg/envoy"
+	envoyAPI "github.com/cilium/cilium/pkg/envoy/cilium"
+	"github.com/cilium/cilium/pkg/envoy/xds"
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/kvstore"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// DefaultAddressSpace is the address space used if none is provided.
+	DefaultAddressSpace = "default"
+)
+
+var (
+	// IPIdentitiesPath is the path to where endpoint IPs are stored in the key-value
+	//store.
+	IPIdentitiesPath = path.Join(kvstore.BaseKeyPrefix, "state", "endpointIPs", "v1")
+
+	// IPIdentityCache caches the mapping of endpoint IPs to their corresponding
+	// security identities across the entire cluster in which this instance of
+	// Cilium is running.
+	IPIdentityCache = NewIPCache(envoy.NetworkPolicyHostsCache)
```

Comment:
I don't like the dependency on the envoy package. It's ok short term but we should factor out the IP to identity cache into a common package in a non XDS specific way.

Reply:
It could be factored out into the watcher instead. Then the two would be decoupled. I'd do a lookup, sort into a list, and then Upsert into the NetworkPolicyHostsCache.